Grape formatter feature requested in #1258 - Rebased and Repushed (#1273)

johnhamelink opened the pull request:
I've rebased, squashed and repushed #1273 so we can get this merged!

 - adds handling for when the returned resource is not serializable via ams
 - fix for when resource is an Array
 - Moves grape include to grape namespace. Changes Enumerable to Array because a plain hash is enumerable.
 - Add integration test
 - Refine scope of Grape version dependency
 - Assert that the response is equal to a manually defined JSON string
 - Add single module to include in Grape projects
 - Create a Serializable Resource to test rails-api from Grape
 - Update docs
 - Fix discrepency between ActiveRecord 4.0 - 4.1 and 4.2
 - Remove different behaviour for Arrays
 - Clean up copy in documentation

**Edit:** Just want to explicitly point out that credit for code goes to Julian Paas (https://github.com/jpaas)

## Todos:

- [x] Add instructions to docs
- [x] Use `self.call` instead of a singleton
- [x] Remove `.gitignore`
- [x] Add Unit tests
- [x] Add single module for including the formatter & helper in their Grape project
- [x] Assert JSON response & not the ability for the JSON parser to parse it (see [here](https://github.com/rails-api/active_model_serializers/pull/1336/files#r45619253))
- [x] Fix tests for Rails < 4.1
- [x] Squash & Rebase

bf4 commented on lib/grape/formatters/active_model_serializers.rb:
```diff
@@ -0,0 +1,18 @@
+  module Formatters
+    module ActiveModelSerializers
+      class << self
+        def call(resource, env)
```
Look great.  

1. would you mind changing this to `def self.call(resource, env)`?  I don't think there's any advantage to opening the singleton class here
1. could you add the documentation from `lib/grape/active_model_serializers.rb` to the docs or README as you see fit.  (I think there's an integrations page in progress somewhere)
1. Could you add a test so we can better support this code?

bf4 replied:
ref: https://github.com/rails-api/active_model_serializers/pull/1312

johnhamelink replied:
@bf4 I'll get that done for you, hopefully within the next 48 hours.

bf4 replied:
Fantastic. I apologize for not bring it up earlier.  Somehow didn't see it

beauby commented on .gitignore:
```diff
@@ -21,3 +21,4 @@ tmp
 *.swp
 .ruby-version
 .ruby-gemset
+.idea
```
Non project-specific ignores do not belong in the `.gitignore`. You should pur them in your local ignore file.

bf4 replied:
True enough, but there's already non-project stuff in there 

B mobile phone

> On Nov 19, 2015, at 5:47 PM, Lucas Hosseini <notifications@github.com> wrote:
> 
> In .gitignore:
> 
> > @@ -21,3 +21,4 @@ tmp
> >  *.swp
> >  .ruby-version
> >  .ruby-gemset
> > +.idea
> Non project-specific ignores do not belong in the .gitignore. You should pur them in your local ignore file.
> 
> —
> Reply to this email directly or view it on GitHub.
>

beauby replied:
True enough, but not reason to add more imho.

johnhamelink commented:
## Todos:

- [x] Add instructions to docs
- [x] Use `self.call` instead of a singleton
- [x] Remove `.gitignore`
- [ ] Add Unit tests

johnhamelink commented:
@bf4 wrt to building a test, since it's not my code and I'm not familiar with how AMS works internally, I'm struggling to come up with a good strategy for testing effectively - could you give me some pointers so I don't waste time on a crappy test?

bf4 commented:
Maybe take a look at how grape or a grape formatter tests it?  You'd likely have to add grape as a dev dependency

B mobile phone

> On Nov 19, 2015, at 7:16 PM, John Hamelink <notifications@github.com> wrote:
> 
> @bf4 wrt to building a test, since it's not my code and I'm not familiar with how AMS works internally, I'm struggling to come up with a good strategy for testing effectively - could you give me some pointers so I don't waste time on a crappy test?
> 
> —
> Reply to this email directly or view it on GitHub.
>

johnhamelink commented:
@bf4 I've implemented some tests which show that it works with Grape - I added one to show that adding metadata works, but the test fails on all Rails versions except for 4.1 - Could you provide insight as to why that might be? Is this a bug in Grape or in this PR?

bf4 commented on active_model_serializers.gemspec:
```diff
@@ -47,4 +47,5 @@ Gem::Specification.new do |spec|
 
   spec.add_development_dependency 'bundler', '~> 1.6'
   spec.add_development_dependency 'timecop', '~> 0.7'
+  spec.add_development_dependency 'grape', '~> 0.13.0'
```
Just curious, do you mean to be `0.13.x` where `x >= 0` or `0.x` where `x >= 13`?. If the latter, it should be `~> 0.13`

johnhamelink replied:
The latter. Thanks for pointing that out :smile:

bf4 commented on docs/howto/grape.md:
```diff
@@ -0,0 +1,27 @@
+    helpers Grape::Helpers::ActiveModelSerializers
+  end
+
+end
```
we can make that mixin here, so that the instructions are `include  Grape::ActiveModelSerializers` or something like that. Thoughts?

bf4 replied:
That would remove the need for the comments in lib/grape/active_model_serializers.rb

johnhamelink replied:
I can't think of any objective reason why not - if a user want's to include the formatter & helpers separately or in their own module then they're still free to do so. Will add to the TODO list.

bf4 commented on test/grape_test.rb:
```diff
@@ -0,0 +1,49 @@
+  def test_formatter_returns_json
+    get '/grape/render'
+    assert last_response.ok?
+    assert JSON.parse(last_response.body)
```
it would be better to assert the expected json response

johnhamelink replied:
:+1:

bf4 commented on test/grape_test.rb:
```diff
@@ -0,0 +1,49 @@
+
+    post = ARModels::Post.new(title: 'Dummy Title', body: 'Lorem Ipsum')
+    serializer = ARModels::PostSerializer.new(post)
+    adapter = ActiveModel::Serializer::Adapter::JsonApi.new(serializer, meta: { page: 1, total_pages: 2 })
```
This should be

```ruby
serializable_resource = ActiveModel::SerializableResource.new(post, serializer: ARModels::PostSerializer, adapter: :json_api, meta: { page: 1, total_pages: 2})`
assert last_response.ok?
assert_equal serializable_resource.to_json, last_response.body
```

or, using the `serializable` test helper which is the same, but less verbose https://github.com/rails-api/active_model_serializers/blob/master/test/support/serialization_testing.rb#L34-L35

and follow the same pattern in https://github.com/rails-api/active_model_serializers/pull/1336/files#r45619253

johnhamelink replied:
Thank you! I will implement this right away :)

bf4 commented on lib/grape/active_model_serializers.rb:
```diff
@@ -0,0 +1,15 @@
+# To add grape support, require 'grape/active_model_serializers' in the base of your grape endpoints
+# Then add 'formatter :json, Grape::Formatters::ActiveModelSerializers' to the endpoints
+# Then add 'helpers Grape::Helpers::ActiveModelSerializers' to the endpoints
```
out of date comments :)  can be condensed to 

```plain
# Mixin for Grape
# Usage: include `Grape::ActiveModelSerializers` in your `Grape::API` endpoint
```

johnhamelink replied:
Fixed :+1:

bf4 commented on docs/howto/grape.md:
```diff
@@ -0,0 +1,13 @@
+# Integration with Grape
+
+To add Grape support, `require 'grape/active_model_serializers'` in the base of your Grape endpoints like so:
```
Let's link to `Grape` repo, and maybe mention supported versions of grape are specified in the gemspec.

I think it should also reference the configuration how-to

johnhamelink replied:
Done :+1:

bf4 commented on lib/grape/formatters/active_model_serializers.rb:
```diff
@@ -0,0 +1,16 @@
+        return resource.to_json unless resource.is_a?(Array) || resource.respond_to?(:model_name)
+        serializer_options = {}
+        serializer_options.merge!(env[:active_model_serializer_options]) if env[:active_model_serializer_options]
+        ActiveModel::SerializableResource.new(resource, serializer_options).to_json
```
naive question. Does grape require the formatter to return a JSON string?

johnhamelink replied:
Grape can handle Hashes, Strings, Arrays, and Enumerables - I will add an update to return the serializable hash

bf4 replied:
really should be `as_json` (not serializable_hash).  and possibly pass some options into it like include, etc.  But I don't know what grape options have at this point

johnhamelink replied:
Apparently I was wrong about this - if I use `as_json` the response is a ruby hash

bf4 commented on lib/grape/helpers/active_model_serializers.rb:
```diff
@@ -0,0 +1,16 @@
+module Grape
+  module Helpers
+    module ActiveModelSerializers
+      # A convenience method for passing ActiveModelSerializer serializer options
```
typo: ActiveModelSerializer**s**

johnhamelink replied:
Fixed

bf4 commented:
Thanks for all your work on this.  I'm hoping you can figure out the build failures so I don't have to... a quick look makes it seem it's not using the json api adapter.

johnhamelink commented:
@bf4 Figured it out - there's a difference between how ActiveRecord 4.2 and 4.0 - 4.1 handle `resource.respond_to?(:model_name)`. I added a check for both `resource.respond_to?(:model_name)` and `resource.class.respond_to?(:model_name)` and that cleaned things up.

bf4 commented on CHANGELOG.md:
```diff
@@ -14,6 +14,7 @@ Breaking changes:
 
 Features:
 
+- [#1336](https://github.com/rails-api/active_model_serializers/pull/1336) Added support for Grape 0.13.x
```
should be `>= 0.13, < 1.0`

bf4 replied:
per `+  spec.add_development_dependency 'grape', '~> 0.13'`

johnhamelink replied:
Done

bf4 commented on docs/howto/grape.md:
```diff
@@ -0,0 +1,19 @@
+
+[Grape](https://github.com/ruby-grape/grape) is an opinionated micro-framework for creating REST-like APIs in ruby.
+
+ActiveModelSerializers currently supports Grape 0.13.x
```
ref: https://github.com/rails-api/active_model_serializers/pull/1336/files#r45766891

johnhamelink replied:
Done

bf4 commented on docs/howto/grape.md:
```diff
@@ -0,0 +1,19 @@
+
+ActiveModelSerializers currently supports Grape 0.13.x
+
+To add [Grape](https://github.com/ruby-grape/grape) support, add the following in the base of your Grape endpoints to enable the formatter and helper functions:
```
maybe replace `add the following` with something like:

> enable the formatter and helper functions by including `Grape::ActiveModelSerializers` in your base endpoint. For example:

johnhamelink replied:
Done

bf4 commented on docs/howto/grape.md:
````diff
@@ -0,0 +1,19 @@
+end
+```
+
+Aside from this, [configuration](../general/configuration_options.md) of AMS is exactly the same.
````
s/AMS/ActiveModelSerializers

johnhamelink replied:
Fixed

bf4 commented on lib/grape/formatters/active_model_serializers.rb:
```diff
@@ -0,0 +1,16 @@
+  module Formatters
+    module ActiveModelSerializers
+      def self.call(resource, env)
+        return resource.to_json unless resource.is_a?(Array) || resource.respond_to?(:model_name) || resource.class.respond_to?(:model_name)
```
just thinking out loud: might be easier to read as

```ruby
if resource.is_a...
  resource.to_json
else
  ..new(..).to_json
end
```

Naive question: ActiveModelSerializers should handle Arrays and non-models just fine. See the logic in [ActionController::Serialization](https://github.com/rails-api/active_model_serializers/blob/5d0cc5990a18c9371441bad90cd8c50f642640fd/lib/action_controller/serialization.rb#L29-L55).   Any reason you're checking for an Array or `model_name`?

johnhamelink replied:
@bf4 that was put there by the previous developer, I have no idea why it was there, but assumed it was OK due to it being in the previous PR

bf4 replied:
Ok. I just never noticed it till now. Progressive review I guess?

johnhamelink replied:
:smile: better caught now than in upstream master. I'll remove it

johnhamelink commented:
Made the documentation changes mentioned in inline comments, rebased & pushed :smile:

bf4 commented on test/grape_test.rb:
```diff
@@ -0,0 +1,89 @@
+          post,
+          meta: { page: 1, total_pages: 2 },
+          adapter: :json_api
+        )
```
Nice! :100: 

Is the `(` and `)` around render in these two routes but not in the one-liner a grape convention?

johnhamelink replied:
Not to my knowledge! Fixed.

bf4 commented:
@johnhamelink Thanks for being so responsive to my comments. Please push back if there's anything that doesn't jive with you.  Sometimes I worry if everything I write just happens :)

johnhamelink commented:
@bf4 Of course, I will do that as soon as you say something that isn't sensible ^.^

johnhamelink commented:
@bf4 any idea why travis failed this time? Seems to have passed all the tests and failed afterwards? :s

bf4 commented:
@johnhamelink Looks like Grape warnings aren't being ignored https://travis-ci.org/rails-api/active_model_serializers/jobs/92987377

https://github.com/rails-api/active_model_serializers/blob/master/test/capture_warnings.rb#L31 `      line.include?(app_root) && !line.include?(bundle_dir)` needs to also ignore the `vendor_dir` since https://github.com/rails-api/active_model_serializers/pull/1341/files

```plain

    ---------------------- app warnings: ----------------------

    /home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/settings.rb:15: warning: method redefined; discarding old top_level_setting
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/settings.rb:21: warning: method redefined; discarding old inheritable_setting
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/routing.rb:173: warning: method redefined; discarding old routes
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/routing.rb:153: warning: instance variable @namespace_description not initialized
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/routing.rb:154: warning: instance variable @namespace_description not initialized
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/endpoint.rb:20: warning: instance variable @before_each not initialized
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/inside_route.rb:182: warning: instance variable @file not initialized
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/inside_route.rb:166: warning: instance variable @body not initialized
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/inside_route.rb:106: warning: instance variable @status not initialized
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/endpoint.rb:20: warning: instance variable @before_each not initialized
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/inside_route.rb:182: warning: instance variable @file not initialized
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/inside_route.rb:166: warning: instance variable @body not initialized
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/inside_route.rb:106: warning: instance variable @status not initialized
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/endpoint.rb:20: warning: instance variable @before_each not initialized
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/inside_route.rb:182: warning: instance variable @file not initialized
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/inside_route.rb:166: warning: instance variable @body not initialized
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.1.0/gems/grape-0.13.0/lib/grape/dsl/inside_route.rb:106: warning: instance variable @status not initialized

    -----------------------------------------------------------
rake aborted!
```

bf4 commented:
ref: https://github.com/rails-api/active_model_serializers/issues/1258

bf4 commented:
Ok, rebase off of master and they should pass https://github.com/rails-api/active_model_serializers/pull/1349

johnhamelink commented:
@bf4 All good to merge?

bf4 commented:
@johnathanludwig I'm sorry this hasn't been merged.   I'd prefer to get a second LGTM on it, so @rails-api/ams any objects to merging this? 24 hours till merge.

Also, you'll have to rebase it off master and resolve conflicts. Let me know if you'd like help.

johnathanludwig commented:
@bf4 I think you meant to tag @johnhamelink

johnhamelink commented:
@bf4 I've rebased and resolved the conflicts (sensibly I hope). Let me know if you've any further comments!

bf4 commented on active_model_serializers.gemspec:
```diff
@@ -53,4 +53,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency 'bundler', '~> 1.6'
   spec.add_development_dependency 'timecop', '~> 0.7'
   spec.add_development_dependency 'minitest-reporters'
+  spec.add_development_dependency 'grape', '~> 0.13'
```
should be `[' >= 0.13', '< 1.0'`

johnhamelink replied:
Fixed

bf4 commented on lib/grape/formatters/active_model_serializers.rb:
```diff
@@ -0,0 +1,15 @@
+# A grape response formatter that can be used as 'formatter :json, Grape::Formatters::ActiveModelSerializers'
```
super small nitpick: If you're already making changes, would be good to capitalize `grape` here and elsewhere

johnhamelink replied:
Fixed :)

bf4 commented on test/grape_test.rb:
```diff
@@ -0,0 +1,85 @@
+
+      get '/render_with_json_api' do
+        post = ARModels::Post.new(title: 'Dummy Title', body: 'Lorem Ipsum')
+        render(post, meta: { page: 1, total_pages: 2 }, adapter: :json_api)
```
style: parens after render can be removed. if they'er a grape convention, then they should be added on line 14 where missing

johnhamelink replied:
Fixed

bf4 commented on test/grape_test.rb:
```diff
@@ -0,0 +1,85 @@
+    get '/grape/render_with_json_api'
+
+    post = ARModels::Post.new(title: 'Dummy Title', body: 'Lorem Ipsum')
+    serializable_resource = serializable(post, serializer: ARModels::PostSerializer, adapter: :json_api, meta: { page: 1, total_pages: 2 })
```
I'd just like to note that the other two tests here have an `expected` hash but this one generates the expected hash.   I'm not sure if I like one style over the other or if there's a reason they're different. So, just noting it in case you have any thoughts.

johnhamelink replied:
Good point - I've moved towards using `serializable` everywhere. I also added `assert last_response.ok?` to the one test that was missing it.

bf4 commented on docs/README.md:
```diff
@@ -23,7 +23,7 @@ This is the documentation of AMS, it's focused on the **0.10.x version.**
 | Ember.js | 0.9.x | [active-model-adapter](https://github.com/ember-data/active-model-adapter)
 | Ember.js | 0.10.x + |  [docs/integrations/ember-and-json-api.md](integrations/ember-and-json-api.md)
-| Grape | 0.10.x + | [#1258](https://github.com/rails-api/active_model_serializers/issues/1258)  |
+| Grape | 0.10.x + | [docs/integrations/grape.md](integrations/grape.md)  |
```
awesome! Thanks for editing this and the other doc files!

bf4 commented on test/grape_test.rb:
```diff
@@ -0,0 +1,89 @@
+  def test_formatter_handles_arrays
+    get '/grape/render_array_with_json_api'
+
+    expected = {
```
serializable? or I can merge it

johnhamelink replied:
How could I create an array to pass into the serializer?

bf4 replied:
If you modified the routing as

```diff
+  def setup
+    @post = ARModels::Post.create(title: 'Dummy Title', body: 'Lorem Ipsum')
+     ARModels::Post.create(title: 'post 2', body: 'body 2')
+     @posts = ARModels::Post.all
+  end
    resources :grape do
      get '/render' do
-        post = ARModels::Post.new(title: 'Dummy Title', body: 'Lorem Ipsum')
+        render @post
      end

      get '/render_with_json_api' do
-       post = ARModels::Post.new(title: 'Dummy Title', body: 'Lorem Ipsum')
-        render post, meta: { page: 1, total_pages: 2 }, adapter: :json_api
+        render @post, meta: { page: 1, total_pages: 2 }, adapter: :json_api
      end

      get '/render_array_with_json_api' do
-        render ARModels::Post.all, adapter: :json_api
+        render @posts, adapter: :json_api
      end
    end
```

Then your tests become

bf4 replied:
thoughts

johnhamelink replied:
Sorry I never got back to you on this - work got in the way :smile: 

I tried to implement what you said, but unfortunately it doesn't work due to the fact that `@post` can't be associated from within `GrapeTest` class. I say LGTM.

bf4 commented:
@johnhamelink merged, yay!

johnhamelink commented:
:tada: hurray!

Thanks @bf4 for your great advice & guidance :)
